[workflow] Add Phi-4 Finetuning Example

rajithkrishnegowda opened the pull request:
This PR adds Phi-4 Finetuning example using Workflow API.

rajithkrishnegowda committed:
Adding phi-4 model

rajithkrishnegowda committed:
Adding phi-4 model

rajithkrishnegowda committed:
changes

rajithkrishnegowda committed:
changes

rajithkrishnegowda committed:
added openfl pic

rajithkrishnegowda committed:
added openfl pic

rajithkrishnegowda committed:
changes

rajithkrishnegowda committed:
changes

rajithkrishnegowda committed:
data

rajithkrishnegowda committed:
data

rajithkrishnegowda committed:
adding docstring

rajithkrishnegowda committed:
adding docstring

rajithkrishnegowda committed:
upload pic

rajithkrishnegowda committed:
upload pic

rajithkrishnegowda committed:
add pic

rajithkrishnegowda committed:
add pic

rajithkrishnegowda committed:
remove png

rajithkrishnegowda committed:
remove png

rajithkrishnegowda committed:
adding pic

rajithkrishnegowda committed:
adding pic

rahulga1 commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,798 @@
+   "id": "8b4ff024-7700-4f98-9d18-0224665f873d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
you should have both ways, either directly installing openfl by pip install openfl or cloning repo.

rajithkrishnegowda replied:
Throughout the repo git clone is used. if we do pip install openfl, it won't install workflow_interface_requirements.txt packages

MasterSkepticista replied:
This is not the ideal approach. We want to redirect all install instructions to our docs. In order to install workflow requirements, one has to run `fx experimental activate`.

rajithkrishnegowda replied:
changes done in 0f8a41b22c2ffe8b8209b6f102d4268195fc781a

rahulga1 commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,798 @@
+   "cell_type": "markdown",
+   "id": "812cfcc8-33ec-4a2b-8a74-27bfc2a41d7b",
+   "metadata": {},
+   "source": [
```
I dont think this comment is out of place, as we are not importing anything.

rajithkrishnegowda replied:
changes done in e6e5b006edf28ac65abaef9151d7820f674680ee

rahulga1 commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,798 @@
+   "id": "e120a656-f4a5-47a5-a3d4-62c5f3672bba",
+   "metadata": {},
+   "source": [
+    "You'll notice in the `FederatedFlow` definition above that there were certain attributes that the flow was not initialized with, namely the `train_dataset` and `eval_dataset` for each of the collaborators. These are **private_attributes** that are exposed only throught the runtime. Each participant has it's own set of private attributes: a dictionary where the key is the attribute name, and  ...
```
typo `throught`

rajithkrishnegowda replied:
resolved in 2ba5c9f9d2b3e70a1de0069d9648c0fb13520251

rahulga1 commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,798 @@
+   "id": "a714d927",
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
formating issue.

rajithkrishnegowda replied:
changes done in e6e5b006edf28ac65abaef9151d7820f674680ee

porteratzo commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,798 @@
+    "            exclude=[\"model\"],\n",
+    "        )\n",
+    "\n",
+    "    def launch_gradio(self):\n",
```
This function is never called, is it needed?

rajithkrishnegowda replied:
resolved in 2ba5c9f9d2b3e70a1de0069d9648c0fb13520251

rajithkrishnegowda committed:
changes done

rajithkrishnegowda committed:
Merge branch 'add_phi-4-model' of https://github.com/rajithkrishnegowda/openfl into add_phi-4-model

rajithkrishnegowda committed:
changes done

rajithkrishnegowda committed:
Merge branch 'develop' into add_phi-4-model

psfoley commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,782 @@
+   "source": [
+    "You'll notice in the `FederatedFlow` definition above that there were certain attributes that the flow was not initialized with, namely the `train_dataset` and `eval_dataset` for each of the collaborators. These are **private_attributes** that are exposed only through the runtime. Each participant has its own set of private attributes: a dictionary where the key is the attribute name, and th ...
+    "\n",
+    "Below, we segment shards of the MedQuAD dataset for **three collaborators**: Portland, Seattle, and Chandler. Each has their own slice of the dataset that's accessible via the `train_dataset` or `eval_dataset` attribute. Note that the private attributes are flexible, and you can choose to pass in a completely different type of object to any of the collaborators or aggregator (with an arbitra ...
```
This description refers to the wrong dataset and # of collaborators.

rajithkrishnegowda replied:
resolved in 33afdefa7662c4e8a29d7465f86709c8daf33b5a

rajithkrishnegowda committed:
changes done

tanwarsh commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,782 @@
+    "if not os.path.exists(\"math_10k.json\"):\n",
+    "    r = requests.get(\n",
+    "        \"https://raw.githubusercontent.com/AGI-Edgerunners/LLM-Adapters/main/ft-training_set/math_10k.json\",\n",
+    "        # \"math_10k.json\", timeout=10\n",
```
NIT: Remove comment

rajithkrishnegowda replied:
changes done in 60e7b0c5860e9ce8a15d69fe444c124cbfd42499

tanwarsh commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,782 @@
+    "tokenizer = AutoTokenizer.from_pretrained(checkpoint_path)\n",
+    "sequence_max_length = 512\n",
+    "val_set_size = 2000\n",
+    "tokenizer.pad_token_id = 0  # unk. we want this to be different from the eos token\n",
```
Typo? unk?

rajithkrishnegowda replied:
changes done in 60e7b0c5860e9ce8a15d69fe444c124cbfd42499

tanwarsh commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,782 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import gradio as gr\n",
```
This is already imported in the start of notebook!

rajithkrishnegowda replied:
changes done in 60e7b0c5860e9ce8a15d69fe444c124cbfd42499

rajithkrishnegowda committed:
changes

MasterSkepticista reviewed (changes requested):
* Why exactly do we need Gradio? How is it related to an FL tutorial?
* The PR description/title is misleading. Please rename as "[workflow] Add Phi-4 Finetuning Example"
* Consider asking the user to run `fx experimental activate` after installing OpenFL. This need not be part of the notebook execution cells.

rajithkrishnegowda committed:
added a section to install openfl

rajithkrishnegowda commented:
> * Why exactly do we need Gradio? How is it related to an FL tutorial?
> * The PR description/title is misleading. Please rename as "[workflow] Add Phi-4 Finetuning Example"
> * Consider asking the user to run `fx experimental activate` after installing OpenFL. This need not be part of the notebook execution cells.

1. Gradio is a user-friendly interface that simplifies the interaction with machine learning models. I have added a Federated Learning (FL) tutorial that allows users to input and submit the number of rounds for federated learning. This feature enhances interactivity and helps users better understand the iterative nature of federated learning.
2. Changed the PR description/title
3. added a note section for user to install openfl iv `pip install -U openfl` and to run `fx experimental activate`

payalcha commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,789 @@
+   "outputs": [],
+   "source": [
+    "!pip install git+https://github.com/intel/openfl.git --no-cache\n",
+    "!cd PATH_TO_OPENFL_DIR/openfl-tutorials/experimental/workflow && pip install -r workflow_interface_requirements.txt --no-cache"
```
Shouldn't we use 
#!pip install -r https://raw.githubusercontent.com/intel/openfl/develop/openfl-tutorials/experimental/workflow/workflow_interface_requirements.txt 
instead of cd.
It will maintain similarity in different workbooks.

rajithkrishnegowda replied:
changes made as per karan suggestions

payalcha commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,789 @@
+    "        \"\"\"\n",
+    "        print(f\"Performing initialization for model\")\n",
+    "        self.collaborators = self.runtime.collaborators\n",
+    "        self.private = 10\n",
```
is this needed.

rajithkrishnegowda replied:
resolved in 6ebb95e051361ae47b7813806ad20fa49d047eb5 
removed as it is not used

payalcha commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,789 @@
+    "        self.next(\n",
+    "            self.aggregated_model_validation,\n",
+    "            foreach=\"collaborators\",\n",
+    "            exclude=[\"model\"],\n",
```
model variable is not there in start of aggregator, do we need to exclude model here?

rajithkrishnegowda replied:
resolved in 6ebb95e051361ae47b7813806ad20fa49d047eb5 
The `model` is not being modified or used in the start method hence removed it

MasterSkepticista commented:
> 1. I have added a Federated Learning (FL) tutorial that allows users to input and submit the number of rounds for federated learning. This feature enhances interactivity and helps users better understand the iterative nature of federated learning.

* For (1), my recommendation is to remove gradio and the interactive element (if this existed for older notebooks, I suggest updating them too). A notebook interface is interactive enough. IMO adding GUI is out-of-place for a fine-tuning example. Furthermore, we really want to reduce the amount of version maintenance we do.
* For (3), my suggestion was to actually not install `openfl` in the notebook. It is expected that the user has already installed OpenFL by the time they clone the repository. It is perhaps the first thing they are expected to do. In this (and all other) notebooks, a markdown cell can point the user to installation page on official docs _and_ then the step to activate experimental APIs. Installing any packages specific to the notebook is fine, though.

psfoley commented:
> > 1. I have added a Federated Learning (FL) tutorial that allows users to input and submit the number of rounds for federated learning. This feature enhances interactivity and helps users better understand the iterative nature of federated learning.
> 
> * For (1), my recommendation is to remove gradio and the interactive element (if this existed for older notebooks, I suggest updating them too). A notebook interface is interactive enough. IMO adding GUI is out-of-place for a fine-tuning example. Furthermore, we really want to reduce the amount of version maintenance we do.
> * For (3), my suggestion was to actually not install `openfl` in the notebook. It is expected that the user has already installed OpenFL by the time they clone the repository. It is perhaps the first thing they are expected to do. In this (and all other) notebooks, a markdown cell can point the user to installation page on official docs _and_ then the step to activate experimental APIs. Installing any packages specific to the notebook is fine, though.

1. Agree on this point. I'm personally a fan of gradio - and it would be useful to show in potentially other dedicated notebooks how other higher level interfaces can be built on top of the Workflow API, but in this notebook we want the focus to be LLM usage.
2. OpenFL indeed should be installed ahead of time. `workflow_interface_requirements.txt` is only kept as separate for now because this is an experimental package. These will eventually be merged into the requirements of setup.py once the experimental label for the Workflow API is dropped. Installing packages needed for a particular notebook example is standard practice.

rajithkrishnegowda committed:
code changes

rajithkrishnegowda commented:
> > 1. I have added a Federated Learning (FL) tutorial that allows users to input and submit the number of rounds for federated learning. This feature enhances interactivity and helps users better understand the iterative nature of federated learning.
> 
> * For (1), my recommendation is to remove gradio and the interactive element (if this existed for older notebooks, I suggest updating them too). A notebook interface is interactive enough. IMO adding GUI is out-of-place for a fine-tuning example. Furthermore, we really want to reduce the amount of version maintenance we do.
> * For (3), my suggestion was to actually not install `openfl` in the notebook. It is expected that the user has already installed OpenFL by the time they clone the repository. It is perhaps the first thing they are expected to do. In this (and all other) notebooks, a markdown cell can point the user to installation page on official docs _and_ then the step to activate experimental APIs. Installing any packages specific to the notebook is fine, though.

changes made in 6ebb95e051361ae47b7813806ad20fa49d047eb5

MasterSkepticista reviewed (changes requested):
Almost there. A few polish tips:
1. Heading styles inconsistent across the notebook. Suggest using h1 for title (i.e. single # in markdown) and h2 for sections, h3 for sub-sections. Note that I don't mean to enforce using h1-h2-h3. Some notebooks are also good with just h1 and h3.
2. Update PR description (since now not using gradio)
3. Re-run the notebook and include cell outputs. As mentioned in previous notebook PRs, having cell outputs is useful for the _user_. This helps a user guess whether the notebook is worth their time investment very quickly.

rahulga1 commented:
> Almost there. A few polish tips:
> 
> 1. Heading styles inconsistent across the notebook. Suggest using h1 for title (i.e. single # in markdown) and h2 for sections, h3 for sub-sections. Note that I don't mean to enforce using h1-h2-h3. Some notebooks are also good with just h1 and h3.
> 2. Update PR description (since now not using gradio)
> 3. Re-run the notebook and include cell outputs. As mentioned in previous notebook PRs, having cell outputs is useful for the _user_. This helps a user guess whether the notebook is worth their time investment very quickly.

@MasterSkepticista I have a different view on keeping the cell output in notebook. It hampers the readability, these examples are there for reference purpose to demonstrate capability. 
Anyways we are in process of removing non-essential examples out of this repo and the one which we keep will be in included in CI, to ensure new changes are not breaking examples.

rajithkrishnegowda committed:
updating Heading styles consistent across notebook

rajithkrishnegowda commented:
> Almost there. A few polish tips:
> 
> 1. Heading styles inconsistent across the notebook. Suggest using h1 for title (i.e. single # in markdown) and h2 for sections, h3 for sub-sections. Note that I don't mean to enforce using h1-h2-h3. Some notebooks are also good with just h1 and h3.
> 2. Update PR description (since now not using gradio)
> 3. Re-run the notebook and include cell outputs. As mentioned in previous notebook PRs, having cell outputs is useful for the _user_. This helps a user guess whether the notebook is worth their time investment very quickly.

1. changes made in f785846f8079818aed18519f0f31522136809081 
2. updated PR description

rajithkrishnegowda committed:
Merge branch 'develop' into add_phi-4-model

kminhta commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,716 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "training_config = {\n",
```
I recommend cleaning up these parameters

- remove parameters that are already set to their default values (for example, `use_ipex`, `logging_strategy`, etc.). It will improve readability and reduce cognitive overload. 

- it would be helpful to call out anything was specifically set to enable federated fine-tuning (i.e. setting 'num_train_epochs' to 1)

rajithkrishnegowda replied:
Changes done in e10dadc7c8ccbeab81f719e02cd20860c7f5ce62

kminhta commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,716 @@
+    "    \"warmup_ratio\": 0.2,\n",
+    "}\n",
+    "\n",
+    "peft_config = {\n",
```
Same here, would be good to clean it up

rajithkrishnegowda replied:
The default parameters are not set in peft config.

kminhta commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,716 @@
+   "metadata": {},
+   "source": [
+    "After installation, activate experimental APIs using:   \n",
+    "`fx experimental activate`"
```
Question: since we are importing directly from `openfl.experimental` and are not using any `fx` commands, do we need to explicitly activate it?

rajithkrishnegowda replied:
Karan suggested that in order to install workflow requirements, one has to run fx experimental activate

kminhta reviewed (commented):
Looking good @rajithkrishnegowda, just some minor questions/comments

MasterSkepticista commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,716 @@
+   "source": [
+    "## Installing OpenFL\n",
+    "To install OpenFL, follow the official documentation: \n",
+    "[OpenFL Installation Guide](https://openfl.readthedocs.io/en/v1.6/get_started/installation.html)"
```
This is pointing to the old doc. Consider pointing it to `latest`.

rajithkrishnegowda replied:
changes done in 711112c1528a3f676be13ab6f626442a63ac7699

rajithkrishnegowda committed:
updated installation guide link

rajithkrishnegowda committed:
removed default values

rajithkrishnegowda committed:
Merge branch 'develop' into add_phi-4-model

MasterSkepticista commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,711 @@
+   "source": [
+    "## 🎉 Congratulations! 🎉\n",
+    "\n",
+    "Now that you've completed this notebook, check out our [other tutorials](https://github.com/securefederatedai/openfl/tree/886704508b8b3b0638372003d72e0bcf7f2e7114/openfl-tutorials/experimental)\n",
```
This URL is over a year old. Replace with `develop` URL.

rajithkrishnegowda replied:
changes done in 1a1084f04db393203b5367e7f223307a22f01d3b

MasterSkepticista commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,711 @@
+  },
+  {
+   "attachments": {
+    "8c48fe4d-2bd6-4f5e-94dd-22dc904d4f62.png": {
```
Can we point to a static image from OpenFL documentation (or might add one under `docs/images` if not exist)? Inline image bytes bloat the size of a notebook.

rajithkrishnegowda replied:
resolved in 0c28026e8abbda5d19fd737485d6034d69151600

MasterSkepticista commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,711 @@
+   "outputs": [],
+   "source": [
+    "# Install dependencies \n",
+    "!pip install torch transformers peft datasets trl==0.12.2 gradio -q"
```
Forgot to remove `gradio`?

rajithkrishnegowda replied:
Changes in 1a1084f04db393203b5367e7f223307a22f01d3b

MasterSkepticista commented on openfl-tutorials/experimental/workflow/LLM/phi-4.ipynb:
```diff
@@ -0,0 +1,711 @@
+    "from transformers import AutoModelForCausalLM, AutoTokenizer, TrainingArguments\n",
+    "from transformers.trainer_callback import PrinterCallback\n",
+    "from trl import SFTTrainer\n",
+    "import gradio as gr\n",
```
And its imports?

rajithkrishnegowda replied:
changes done in 1a1084f04db393203b5367e7f223307a22f01d3b

rajithkrishnegowda committed:
code changes

rajithkrishnegowda committed:
corrected path for workflow image

rajithkrishnegowda committed:
updated workflow image

MasterSkepticista commented:
Thanks for the updates @rajithkrishnegowda , it looks polished now!

kminhta reviewed (approved):
thanks for the changes, LGTM!
